pkg/services/ngalert/notifier/receivers.go

```diff
@@ -0,0 +1,214 @@
+}
+
+func (e ReceiverTimeoutError) Error() string {
+	return fmt.Sprintf("the receiver timed out: %s", e.Err)
+}
+
+func (am *Alertmanager) TestReceivers(ctx context.Context, c apimodels.TestReceiversConfigParams) (*TestReceiversResult, error) {
+	// now represents the start time of the test
+	now := time.Now()
+	testAlert := &types.Alert{
+		Alert: model.Alert{
+			Labels: model.LabelSet{
+				model.LabelName("alertname"): "TestAlertAlwaysFiring",
+				model.LabelName("instance"):  "Grafana",
+			},
+			Annotations: model.LabelSet{},
+			StartsAt:    now,
+		},
+		UpdatedAt: now,
+	}
+
+	// we must set a group key as this is required by some notification channels (i.e. webhook)
+	ctx = notify.WithGroupKey(ctx, testAlert.Labels.String())
+
+	tmpl, err := am.getTemplate()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get template: %w", err)
+	}
+
+	// job contains all metadata required to test a receiver
+	type job struct {
+		Config       *apimodels.PostableGrafanaReceiver
+		ReceiverName string
+		Notifier     notify.Notifier
+	}
+
+	// result contains all metadata required to test a receiver
+	// and an error that is non-nil if the test fails
+	type result struct {
+		Config       *apimodels.PostableGrafanaReceiver
+		ReceiverName string
+		Error        error
+	}
+
+	invalid := make([]result, 0, len(c.Receivers))
+	jobs := make([]job, 0, len(c.Receivers))
+
+	for _, receiver := range c.Receivers {
+		for _, next := range receiver.GrafanaManagedReceivers {
+			n, err := am.buildReceiverIntegration(next, tmpl)
+			if err != nil {
+				invalid = append(invalid, result{
+					Config:       next,
+					ReceiverName: receiver.Name,
+					Error:        err,
+				})
+			} else {
+				jobs = append(jobs, job{
+					Config:       next,
+					ReceiverName: receiver.Name,
+					Notifier:     n,
+				})
+			}
+		}
+	}
+
+	numWorkers := maxTestReceiversWorkers
+	if numWorkers > len(jobs) {
+		numWorkers = len(jobs)
+	}
+
+	resultCh := make(chan result, len(jobs))
+	workCh := make(chan job, len(jobs))
+	for _, job := range jobs {
+		workCh <- job
+	}
+	close(workCh)
+
+	g, ctx := errgroup.WithContext(ctx)
+	for i := 0; i < numWorkers; i++ {
+		g.Go(func() error {
+			for next := range workCh {
+				v := result{
+					Config:       next.Config,
+					ReceiverName: next.ReceiverName,
+				}
+				if _, err := next.Notifier.Notify(ctx, testAlert); err != nil {
+					v.Error = err
+				}
+				resultCh <- v
+			}
+			return nil
+		})
+	}
+	g.Wait() // nolint
+	close(resultCh)
+
+	// m keeps track of the results for each of the receivers
+	m := make(map[string]TestReceiverResult)
+	for _, receiver := range c.Receivers {
+		// set up the result for this receiver
+		m[receiver.Name] = TestReceiverResult{
+			Name:    receiver.Name,
+			Configs: make([]TestReceiverConfigResult, 0, len(receiver.GrafanaManagedReceivers)),
+		}
+	}
+	for _, next := range invalid {
+		v := m[next.ReceiverName]
+		v.Configs = append(v.Configs, TestReceiverConfigResult{
```

Comment:
nit: we could do this append of invalid configs as part of L101

Reply:
Sure! I had separated this so the code would be easier to read (i.e creating the result all happens together). What do think of this in relation to the extra allocation of the `invalid` slice?

Reply:
It just struck my brain because I was expecting the pickup of invalid results at that point, but this is something that's very subjective and mostly a nit - happy to leave it as is.